### PR TITLE
story VSPC-117

### PR DESCRIPTION
**Pull request by prjadhav14**
Development to show a list of Space Names in which the listed images are used as Background Image.

### CODE REVIEW

**Comment by jdamerow** [comment]
Resolve conflicts please

**Comment by jdamerow** [comment]
@prjadhav14 just assign the review to someone else who is working today.

File: vspace/src/main/java/edu/asu/diging/vspace/core/data/SpaceRepository.java [hunk 1]
```diff
@@ -12,9 +12,11 @@
...
 @JaversSpringDataAuditable
 public interface SpaceRepository extends PagingAndSortingRepository<Space, String> {
-    
+
     List<Space> findTop5ByOrderByCreationDateDesc();
-    
+
     List<Space> findAllBySpaceStatus(SpaceStatus spaceStatus);
 
+    List<Space> findAllByImage_Id(String imageId);
```
Comment (kakhilreddy) [inline vspace/src/main/java/edu/asu/diging/vspace/core/data/SpaceRepository.java:20]: Why an' _' for function name?

File: vspace/src/main/webapp/WEB-INF/views/staff/images/imagelist.jsp [hunk 2]
```diff
@@ -82,6 +82,7 @@ $( document ).ready(function() {
 				<tr>
 					<th scope="col" sort-prop="filename"><a href="#">Filename</a></th>
 					<th scope="col" sort-prop="name"><a href="#">Name</a></th>
+                    <th scope="col" sort-prop="name"><a href="#">Space</a></th>
```
Comment (kakhilreddy) [inline vspace/src/main/webapp/WEB-INF/views/staff/images/imagelist.jsp:85]: Indentation is missing

File: vspace/src/main/webapp/WEB-INF/views/staff/images/imagelist.jsp [hunk 3]
```diff
@@ -97,6 +98,11 @@ $( document ).ready(function() {
 								src="<c:url value="/api/image/${image.id}"/>"
 								class="img-thumbnail"> ${image.filename} </a></th>
 						<td>${image.name}</td>
+                        <td>
+                        <c:forEach items="${imageToSpaces[image.id]}" var="space">
+                            ${space.name}<br>
+                        </c:forEach>    
+                        </td>
```
Comment (kakhilreddy) [inline vspace/src/main/webapp/WEB-INF/views/staff/images/imagelist.jsp:105]: Indentation is missing.

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/ListImagesController.java [hunk 4]
```diff
@@ -38,12 +47,20 @@ public String listSpaces(@PathVariable(required = false) String page,
...
         model.addAttribute("currentPageNumber", pageNo);
         model.addAttribute("totalImageCount", imageService.getTotalImageCount());
-        model.addAttribute("images", imageService.getImages(pageNo, sortedBy, order));
+        List<IVSImage> images = imageService.getImages(pageNo, sortedBy, order);
+        LinkedHashMap<String, List<ISpace>> imageToSpaces = new LinkedHashMap<String, List<ISpace>>();
+        for(IVSImage image : images) {
+            List<ISpace> spaces = spaceManager.getSpacesWithImageId(image.getId());
+            if(spaces.size()>0)
+                imageToSpaces.put(image.getId(), spaces);
+        }
```
Comment (kakhilreddy) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/ListImagesController.java:56]: Can you write unit test cases for it?

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/ListImagesController.java [hunk 5]
```diff
@@ -38,12 +47,20 @@ public String listSpaces(@PathVariable(required = false) String page,
         model.addAttribute("totalPages", imageService.getTotalPages());
         model.addAttribute("currentPageNumber", pageNo);
         model.addAttribute("totalImageCount", imageService.getTotalImageCount());
-        model.addAttribute("images", imageService.getImages(pageNo, sortedBy, order));
+        List<IVSImage> images = imageService.getImages(pageNo, sortedBy, order);
+        LinkedHashMap<String, List<ISpace>> imageToSpaces = new LinkedHashMap<String, List<ISpace>>();
```
Comment (kakhilreddy) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/ListImagesController.java:51]: Doubt: Why are you using LinkedHashMap?
 Is maintaining order required?

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/ListImagesController.java [hunk 6]
```diff
@@ -38,12 +47,20 @@ public String listSpaces(@PathVariable(required = false) String page,
...
+        for(IVSImage image : images) {
+            List<ISpace> spaces = spaceManager.getSpacesWithImageId(image.getId());
```
Comment (kakhilreddy) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/ListImagesController.java:53]: What type of exceptions can spaceManager.getSpacesWithImageId raise?
image Id null or invalid image id?
How is it handled?

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/SpaceManager.java [hunk 7]
```diff
@@ -248,12 +248,20 @@ public void deleteSpaceById(String id) {
...
     @Override
     public List<SpaceLink> getIncomingLinks(String id) {
 
         return spaceLinkRepo.getLinkedFromSpaces(id);
     }
+
+    @Override
+    public List<ISpace> getSpacesWithImageId(String imageId) {
+        List<ISpace> spaces = new ArrayList<>();
+        spaceRepo.findAllByImage_Id(imageId).forEach(space -> spaces.add(space));
```
Comment (kakhilreddy) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/SpaceManager.java:261]: What if imageId is null or invalid?

File: vspace/src/main/java/edu/asu/diging/vspace/core/data/SpaceRepository.java (hunk 1 above)
Comment (prjadhav14) [reply vspace/src/main/java/edu/asu/diging/vspace/core/data/SpaceRepository.java:20]: Yes, changed

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/ListImagesController.java (hunk 5 above)
Comment (prjadhav14) [reply vspace/src/main/java/edu/asu/diging/vspace/web/staff/ListImagesController.java:51]: Not required, changed to Map, HashMap.

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/ListImagesController.java (hunk 6 above)
Comment (prjadhav14) [reply vspace/src/main/java/edu/asu/diging/vspace/web/staff/ListImagesController.java:53]: It is a referenced primary key and hence cannot take null value. Handled on database table configuration.

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/SpaceManager.java (hunk 7 above)
Comment (prjadhav14) [reply vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/SpaceManager.java:261]: Not a valid case as imageIds are retrieved from ImageRepo and then passed to spaceRepo to get list of spaces using particular imageId.

Comment (jdamerow) [reply vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/SpaceManager.java]: in your case, but what about other use cases?

Comment (prjadhav14) [reply vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/SpaceManager.java]: right, will handle the cases.

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/ListImagesController.java (hunk 6 above)
Comment (kakhilreddy) [reply vspace/src/main/java/edu/asu/diging/vspace/web/staff/ListImagesController.java:54]: But what about invalid image id?

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/ListImagesController.java (hunk 5 above)
Comment (kakhilreddy) [reply vspace/src/main/java/edu/asu/diging/vspace/web/staff/ListImagesController.java]: ok!

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/ListImagesController.java [hunk 8]
```diff
@@ -38,12 +48,20 @@ public String listSpaces(@PathVariable(required = false) String page,
...
         model.addAttribute("currentPageNumber", pageNo);
         model.addAttribute("totalImageCount", imageService.getTotalImageCount());
-        model.addAttribute("images", imageService.getImages(pageNo, sortedBy, order));
+        List<IVSImage> images = imageService.getImages(pageNo, sortedBy, order);
+        Map<String, List<ISpace>> imageToSpaces = new HashMap<String, List<ISpace>>();
+        for(IVSImage image : images) {
+            List<ISpace> spaces = spaceManager.getSpacesWithImageId(image.getId());
+            if(spaces!=null && spaces.size()>0)
+                imageToSpaces.put(image.getId(), spaces);
+        }
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/ListImagesController.java:57]: please follow our style guide

File: vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SpaceManagerTest.java [hunk 9]
```diff
@@ -262,4 +263,21 @@ public void test_getSpacesWithUnPublishedStatus_spaceDoesNotExist() throws Space
         Assert.assertTrue(spaceWithNullStatus.isEmpty());
     }
 
+    @Test
+    public void test_getSpacesWithImageId_whenImageIdIsValid() throws SpaceDoesNotExistException{
```
Comment (jdamerow) [inline vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SpaceManagerTest.java:267]: needs a test for when the id is not valid

Comment (jdamerow) [reply vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SpaceManagerTest.java:267]: and 0 and several spaces for an id

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Comment by jdamerow** [comment]
On the image list page, let's add a comma between the spaces, so it's clear that there are several spaces in some cases, and let's make the space name a link to the space staff page.

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/ListImagesController.java (hunk 8 above)
Comment (jdamerow) [reply vspace/src/main/java/edu/asu/diging/vspace/web/staff/ListImagesController.java:57]: ?

File: vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SpaceManagerTest.java [hunk 10]
```diff
@@ -262,4 +263,34 @@ public void test_getSpacesWithUnPublishedStatus_spaceDoesNotExist() throws Space
...
+        VSImage image = new VSImage();
+        image.setId("IMG001");
+        space.setImage(image);
+        List<Space> spaceList=new ArrayList<>();
+        spaceList.add(space);
+        Mockito.when(spaceRepo.findAllByImageId("IMG001")).thenReturn(spaceList);
+        Optional<VSImage> mockImage = Optional.of(image);
+        Mockito.when(imageRepo.findById(image.getId())).thenReturn(mockImage);
+        List<ISpace> foundSpaceWithImageId= managerToTest.getSpacesWithImageId("IMG001");
+        String actualSpaceIdWithGiveImageId = foundSpaceWithImageId.get(0).getId();
```
Comment (jdamerow) [inline vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SpaceManagerTest.java:279]: it's only a test case, but still, this is an unnecessarily long variable name. what's wrong with `actualId` or `actualSpaceId`? ;)

File: vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SpaceManagerTest.java [hunk 11]
```diff
@@ -262,4 +263,34 @@ public void test_getSpacesWithUnPublishedStatus_spaceDoesNotExist() throws Space
...
+    public void test_getSpacesWithImageId_ImageIdIsInValid() throws SpaceDoesNotExistException{
+        Space space=new Space();
+        space.setId(spaceId1);
+        VSImage image = new VSImage();
+        image.setId("IMG001");
+        space.setImage(image);
+        List<Space> spaceList=new ArrayList<>();
+        Mockito.when(spaceRepo.findAllByImageId("IMG002")).thenReturn(spaceList);
+        Mockito.when(imageRepo.findById(Mockito.anyString())).thenReturn(Optional.empty());
+        Assert.assertNull(managerToTest.getSpacesWithImageId(Mockito.anyString()));
```
Comment (jdamerow) [inline vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SpaceManagerTest.java:293]: `Mockito.anyString` is typically used to match arguments (in the sense of whatever string you get, do this), not as input for a method to test. Just use a string.

File: vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SpaceManagerTest.java [hunk 12]
```diff
@@ -262,4 +263,34 @@ public void test_getSpacesWithUnPublishedStatus_spaceDoesNotExist() throws Space
...
+    }
+
```
Comment (jdamerow) [inline vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SpaceManagerTest.java:295]: needs a test case for when argument is null.

**Review by jdamerow — CHANGES_REQUESTED** [review]

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Comment by jdamerow** [comment]
There should be no blank before the comma:
image

**Comment by jdamerow** [comment]
Make it so, Jenkins.
